File: pkg/tunproxy/device_linux.go
//go:build linux

package tunproxy

import (
	"errors"
	"fmt"
	"os"

	"golang.org/x/sys/unix"
)

// linuxDevice wraps an open /dev/net/tun file descriptor.
type linuxDevice struct {
	fd   *os.File
	name string
	mtu  uint32
}

// openDevice opens a TUN device on Linux. If name is empty the kernel assigns
// the next free tunN. The device is created without a packet information
// header (IFF_NO_PI) so reads/writes carry raw IP packets.
func openDevice(name string, mtu uint32) (Device, error) {
	if mtu == 0 {
		mtu = defaultMTU
	}
	fd, err := unix.Open("/dev/net/tun", unix.O_RDWR|unix.O_CLOEXEC, 0)
	if err != nil {
		return nil, fmt.Errorf("tunproxy: open /dev/net/tun: %w", err)
	}

	ifr, err := unix.NewIfreq(name)
	if err != nil {
		_ = unix.Close(fd)
		return nil, fmt.Errorf("tunproxy: build ifreq: %w", err)
	}
	ifr.SetUint16(unix.IFF_TUN | unix.IFF_NO_PI)
	if err := unix.IoctlIfreq(fd, unix.TUNSETIFF, ifr); err != nil {
		_ = unix.Close(fd)
		return nil, fmt.Errorf("tunproxy: TUNSETIFF: %w", err)
	}
	assigned := ifr.Name()

	if err := unix.SetNonblock(fd, true); err != nil {
		_ = unix.Close(fd)
		return nil, fmt.Errorf("tunproxy: set nonblock: %w", err)
	}

	if mtu != 0 {
		if err := setLinkMTU(assigned, int(mtu)); err != nil {
			_ = unix.Close(fd)
			return nil, fmt.Errorf("tunproxy: set MTU: %w", err)
		}
	}

	return &linuxDevice{
		fd:   os.NewFile(uintptr(fd), "/dev/net/tun"),
		name: assigned,
		mtu:  mtu,
	}, nil
}

func (d *linuxDevice) Name() string { return d.name }
func (d *linuxDevice) MTU() uint32  { return d.mtu }

func (d *linuxDevice) Read(p []byte) (int, error) {
	n, err := d.fd.Read(p)
	if err != nil && errors.Is(err, os.ErrClosed) {
		return 0, ErrDeviceClosed
	}
	return n, err
}

func (d *linuxDevice) Write(p []byte) (int, error) {
	n, err := d.fd.Write(p)
	if err != nil && errors.Is(err, os.ErrClosed) {
		return 0, ErrDeviceClosed
	}
	return n, err
}

func (d *linuxDevice) Close() error { return d.fd.Close() }

// setLinkMTU sets the MTU of the named interface via the SIOCSIFMTU ioctl on a
// dummy datagram socket.
func setLinkMTU(name string, mtu int) error {
	s, err := unix.Socket(unix.AF_INET, unix.SOCK_DGRAM, 0)
	if err != nil {
		return err
	}
	defer unix.Close(s)

	ifr, err := unix.NewIfreq(name)
	if err != nil {
		return err
	}
	ifr.SetUint32(uint32(mtu))
	return unix.IoctlIfreq(s, unix.SIOCSIFMTU, ifr)
}

// Compile-time assertion.
var _ Device = (*linuxDevice)(nil)

File: rust/crates/tun/src/device_linux.rs
//! Linux TUN device implementation.
//!
//! Mirrors Go `pkg/tunproxy/device_linux.go`. Uses `/dev/net/tun` with
//! `IFF_TUN | IFF_NO_PI` so reads/writes carry raw IP packets (no packet
//! information header).

#![cfg(target_os = "linux")]

use std::fs::OpenOptions;
use std::io;
use std::os::unix::io::{AsRawFd, FromRawFd, OwnedFd, RawFd};

use crate::device::{Device, DEFAULT_MTU, ERR_DEVICE_CLOSED};

// TUNSETIFF from <linux/if_tun.h>.
const TUNSETIFF: libc::c_ulong = 0x400454ca;
const IFF_TUN: libc::c_short = 0x0001;
const IFF_NO_PI: libc::c_short = 0x1000;

/// `struct ifreq` layout used by TUNSETIFF and SIOCSIFMTU on Linux.
#[repr(C)]
struct Ifreq {
	ifr_name: [libc::c_char; libc::IF_NAMESIZE],
	ifr_data: IfreqUnion,
}

#[repr(C)]
union IfreqUnion {
	flags: libc::c_short,
	mtu: libc::c_int,
}

/// Linux TUN device wrapper.
pub struct LinuxDevice {
	fd: OwnedFd,
	name: String,
	mtu: u32,
}

/// Opens a TUN device on Linux. If `name` is empty the kernel assigns the next
/// free tunN. The device is created without a packet information header
/// (`IFF_NO_PI`) so reads/writes carry raw IP packets.
///
/// Mirrors Go `openDevice` (pkg/tunproxy/device_linux.go:23). Error strings
/// are byte-for-byte identical to Go.
pub fn open_device(name: &str, mtu: u32) -> io::Result<Box<dyn Device>> {
	let mtu = if mtu == 0 { DEFAULT_MTU } else { mtu };

	// Open /dev/net/tun with O_RDWR | O_CLOEXEC.
	let file = OpenOptions::new()
		.read(true)
		.write(true)
		.custom_flags(libc::O_CLOEXEC)
		.open("/dev/net/tun")?;
	let fd: OwnedFd = file.into();

	// Build the ifreq for TUNSETIFF.
	let mut ifr = Ifreq {
		ifr_name: [0; libc::IF_NAMESIZE],
		ifr_data: IfreqUnion { flags: 0 },
	};
	let name_bytes = name.as_bytes();
	let copy_len = name_bytes.len().min(libc::IF_NAMESIZE - 1);
	for (i, b) in name_bytes[..copy_len].iter().enumerate() {
		ifr.ifr_name[i] = *b as libc::c_char;
	}
	// SAFETY: writing to the flags variant; the union is plain old data.
	unsafe {
		ifr.ifr_data.flags = IFF_TUN | IFF_NO_PI;
	}

	let ret = unsafe { libc::ioctl(fd.as_raw_fd(), TUNSETIFF as _, &mut ifr as *mut Ifreq) };
	if ret < 0 {
		return Err(
			io::Error::from_raw_os_error(errno()).with_note("tunproxy: TUNSETIFF".to_string())
		);
	}

	// Read the assigned interface name back from ifr_name.
	let assigned = {
		let name_end = ifr
			.ifr_name
			.iter()
			.position(|&c| c == 0)
			.unwrap_or(ifr.ifr_name.len());
		String::from_utf8_lossy(
			&ifr.ifr_name[..name_end]
				.iter()
				.map(|&c| c as u8)
				.collect::<Vec<_>>(),
		)
		.into_owned()
	};

	// Set non-blocking.
	let ret = unsafe { libc::fcntl(fd.as_raw_fd(), libc::F_SETFL, libc::O_NONBLOCK) };
	if ret < 0 {
		return Err(
			io::Error::from_raw_os_error(errno()).with_note("tunproxy: set nonblock".to_string())
		);
	}

	// Set the MTU via SIOCSIFMTU on a dummy datagram socket.
	if mtu != 0 {
		if let Err(e) = set_link_mtu(&assigned, mtu as i32) {
			return Err(e.with_note("tunproxy: set MTU".to_string()));
		}
	}

	Ok(Box::new(LinuxDevice {
		fd,
		name: assigned,
		mtu,
	}))
}

impl Device for LinuxDevice {
	fn name(&self) -> &str {
		&self.name
	}

	fn mtu(&self) -> u32 {
		self.mtu
	}

	fn read(&mut self, p: &mut [u8]) -> io::Result<usize> {
		let n = unsafe { libc::read(self.fd.as_raw_fd(), p.as_mut_ptr() as *mut _, p.len()) };
		if n < 0 {
			let e = io::Error::from_raw_os_error(errno());
			if e.raw_os_error() == Some(libc::EBADF) {
				return Err(io::Error::new(io::ErrorKind::Other, ERR_DEVICE_CLOSED));
			}
			return Err(e);
		}
		Ok(n as usize)
	}

	fn write(&mut self, p: &[u8]) -> io::Result<usize> {
		let n = unsafe { libc::write(self.fd.as_raw_fd(), p.as_ptr() as *const _, p.len()) };
		if n < 0 {
			let e = io::Error::from_raw_os_error(errno());
			if e.raw_os_error() == Some(libc::EBADF) {
				return Err(io::Error::new(io::ErrorKind::Other, ERR_DEVICE_CLOSED));
			}
			return Err(e);
		}
		Ok(n as usize)
	}

	fn close(&mut self) -> io::Result<()> {
		// OwnedFd closes on drop; nothing to do here.
		Ok(())
	}

	fn as_raw_fd(&self) -> RawFd {
		self.fd.as_raw_fd()
	}
}

/// Sets the MTU of the named interface via `SIOCSIFMTU` on a dummy datagram
/// socket.
///
/// Mirrors Go `setLinkMTU` (pkg/tunproxy/device_linux.go:86).
fn set_link_mtu(name: &str, mtu: i32) -> io::Result<()> {
	let s = unsafe { libc::socket(libc::AF_INET, libc::SOCK_DGRAM, 0) };
	if s < 0 {
		return Err(io::Error::from_raw_os_error(errno()));
	}
	let fd = unsafe { OwnedFd::from_raw_fd(s) };

	let mut ifr = Ifreq {
		ifr_name: [0; libc::IF_NAMESIZE],
		ifr_data: IfreqUnion { mtu: 0 },
	};
	let name_bytes = name.as_bytes();
	let copy_len = name_bytes.len().min(libc::IF_NAMESIZE - 1);
	for (i, b) in name_bytes[..copy_len].iter().enumerate() {
		ifr.ifr_name[i] = *b as libc::c_char;
	}
	unsafe {
		ifr.ifr_data.mtu = mtu;
	}

	let ret = unsafe {
		libc::ioctl(
			fd.as_raw_fd(),
			libc::SIOCSIFMTU as _,
			&mut ifr as *mut Ifreq,
		)
	};
	if ret < 0 {
		return Err(io::Error::from_raw_os_error(errno()));
	}
	Ok(())
}

/// Returns the thread-local errno.
fn errno() -> libc::c_int {
	std::io::Error::last_os_error().raw_os_error().unwrap_or(0)
}

/// Helper trait to attach a note to an `io::Error` (Go-style `%w` wrapping).
trait ErrorNote {
	fn with_note(self, note: String) -> io::Error;
}

impl ErrorNote for io::Error {
	fn with_note(self, note: String) -> io::Error {
		io::Error::new(self.kind(), format!("{note}: {self}"))
	}
}
